source/shaders/ibl_filtering.frag

```diff
@@ -149,207 +152,251 @@ float D_Charlie(float sheenRoughness, float NdotH)
+    // the distribution (e.g. lambertian uses a cosine importance)
+    if(u_distribution == cLambertian)
+    {
+        // Cosine weighted hemisphere sampling
+        // http://www.pbr-book.org/3ed-2018/Monte_Carlo_Integration/2D_Sampling_with_Multidimensional_Transformations.html#Cosine-WeightedHemisphereSampling
+        cosTheta = sqrt(1.0 - u);
+        sinTheta = sqrt(u); // equivalent to `sqrt(1.0 - cosTheta*cosTheta)`;
+        phi = 2.0 * MATH_PI * v;
+
+        pdf = cosTheta / MATH_PI; // evaluation for solid angle, therefore drop the sinTheta
+    }
+    else if(u_distribution == cGGX)
+    {
+        // specular mapping
+        float alpha = roughness * roughness;
+        cosTheta = sqrt((1.0 - u) / (1.0 + (alpha*alpha - 1.0) * u));
+        sinTheta = sqrt(1.0 - cosTheta*cosTheta);
+        phi = 2.0 * MATH_PI * v;
+    }
+    else if(u_distribution == cCharlie)
+    {
+        // sheen mapping
+        float alpha = roughness * roughness;
+        sinTheta = pow(u, alpha / (2.0*alpha + 1.0));
+        cosTheta = sqrt(1.0 - sinTheta * sinTheta);
+        phi = 2.0 * MATH_PI * v;
+    }
+
+    // transform the hemisphere sample to the normal coordinate frame
+    // i.e. rotate the hemisphere to the normal direction
+    vec3 localSpaceDirection = normalize(vec3(sinTheta * cos(phi), sinTheta * sin(phi), cosTheta));
+    mat3 TBN = generateTBN(N);
+    vec3 direction = TBN * localSpaceDirection;
+
+    if(u_distribution == cGGX || u_distribution == cCharlie)
+    {
+        pdf = PDF(direction, N, roughness);
+    }
+
+    return vec4(direction, pdf);
 }
 
-float PDF(vec3 V, vec3 H, vec3 N, vec3 L, float roughness)
+// https://github.com/google/filament/blob/master/shaders/src/brdf.fs#L136
+float V_Ashikhmin(float NdotL, float NdotV)
 {
-	if(u_distribution == cLambertian)
-	{
-		float NdotL = dot(N, L);
-		return max(NdotL * (1.0 / MATH_PI), 0.0);
-	}
-	else if(u_distribution == cGGX)
-	{
-		float VdotH = dot(V, H);
-		float NdotH = dot(N, H);
-
-		float D = D_GGX(NdotH, roughness);
-		return max(D * NdotH / (4.0 * VdotH), 0.0);
-	}
-	else if(u_distribution == cCharlie)
-	{
-		float VdotH = dot(V, H);
-		float NdotH = dot(N, H);
-
-		float D = D_Charlie(roughness, NdotH);
-		return max(D * NdotH / abs(4.0 * VdotH), 0.0);
-	}
-
-	return 0.f;
+    return clamp(1.0 / (4.0 * (NdotL + NdotV - NdotL * NdotV)), 0.0, 1.0);
+}
+
+// Mipmap Filtered Samples (GPU Gems 3, 20.4)
+// https://developer.nvidia.com/gpugems/gpugems3/part-iii-rendering/chapter-20-gpu-based-importance-sampling
+float computeLod(float pdf)
+{
+    // IBL Baker (Matt Davidson)
+    // https://github.com/derkreature/IBLBaker/blob/65d244546d2e79dd8df18a28efdabcf1f2eb7717/data/shadersD3D11/IblImportanceSamplingDiffuse.fx#L215
+    float solidAngleTexel = 4.0 * MATH_PI / (6.0 * float(u_width) * float(u_sampleCount));
+    float solidAngleSample = 1.0 / (float(u_sampleCount) * pdf);
+    float lod = 0.5 * log2(solidAngleSample / solidAngleTexel);
+
+    lod += u_lodBias;
+
```

Comment:
I would prefer to have this configurable bias outside of the compute function.
With this, line 295 (`lod = u_lodBias;`) becomes more clear: It is not set to the bias, it is **only** the bias.

Reply:
extracted it